### PR TITLE
UAF-3617 Bug: Stack Trace Error when returning order to create Requisition, PO, and Electronic Invoice.

### PR DESCRIPTION
**Pull request by adamkost**
Includes:
1) modified b2bPOResponse.xsd
2) modified b2bPunchOutOrder.xsd
3) modified electronicInvoice.xsd
These modifications were to copy the various simple data types used since ANT is no longer used to build or deploy KFS.

4) overrode the b2b.punch.back.action.forwarding.url property in the institutional-config.properties file to a working value.
5) create the b2b.purchase.order.dtd property in the institutional-config.properties file to contain the path for the PO.dtd file.
6) Implemented the UA KFS3 B2BPurchaseOrderAZSciquestServiceImpl as B2BPurchaseOrderSciquestServiceImpl, and updating the path of the PO.dtd to include the value in b2b.purchase.order.dtd.

This pull request is being created because the original pull request (#196) was accidentally removed from the ua-development branch. The XSD files from the previous pull request have not been modified since that PR.

### CODE REVIEW

File: kfs-purap/src/main/resources/edu/arizona/kfs/module/purap/spring-purap.xml [hunk 1]
```diff
@@ -147,4 +147,17 @@
 
     <bean id="paymentRequestDao" parent="platformAwareDao" class="edu.arizona.kfs.module.purap.document.dataaccess.impl.PaymentRequestDaoOjb"/>
     	
+
+    <bean id="b2bPurchaseOrderService" class="edu.arizona.kfs.module.purap.document.service.impl.B2BPurchaseOrderSciquestServiceImpl">
+        <property name="b2bDao" ref="b2bDao"/>
+        <property name="requisitionService" ref="requisitionService"/>
+        <property name="parameterService" ref="parameterService"/>
```
Comment (leahasullivan) [inline kfs-purap/src/main/resources/edu/arizona/kfs/module/purap/spring-purap.xml]: Why are we injecting these services and the Dao?  These should be inherited from the parent bean if they are being used.

File: kfs-purap/src/main/resources/edu/arizona/kfs/module/purap/spring-purap.xml [hunk 2]
```diff
@@ -147,4 +147,17 @@
...
+
+        <property name="b2bEnvironment" value="${b2b.environment}" />
+        <property name="b2bUserAgent" value="${b2b.user.agent}" />
+        <property name="b2bPurchaseOrderURL" value="${b2b.purchase.order.url}" />
+        <property name="b2bPurchaseOrderIdentity" value="${b2b.purchase.order.identity}" />
+        <property name="b2bPurchaseOrderPassword" value="${b2b.purchase.order.password}" />
```
Comment (leahasullivan) [inline kfs-purap/src/main/resources/edu/arizona/kfs/module/purap/spring-purap.xml]: If we are simply overriding this bean, why are we re-adding these b2b properties here again, they should be inherited from the parent bean in baseline.  Also, if it we are injecting these in KFS6 (spring-purap.xml), we should stick to how it is in 6.

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+import org.kuali.rice.krad.util.ObjectUtils;
+
+import edu.arizona.kfs.module.purap.PurapConstants;
+import edu.arizona.kfs.sys.KFSConstants;
+
+@SuppressWarnings("unused")
+public class B2BPurchaseOrderSciquestServiceImpl extends org.kuali.kfs.module.purap.document.service.impl.B2BPurchaseOrderSciquestServiceImpl {
+    private static org.apache.log4j.Logger LOG = org.apache.log4j.Logger.getLogger(B2BPurchaseOrderSciquestServiceImpl.class);
+
+    private ConfigurationService configurationService;
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: Injecting the b2b property via spring will eliminate the need for this variable and the getter method.  Please remove and see the comments on the spring-purap.xml.

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+    private DateTimeService dateTimeService;
+
+    private DateTimeService getDateTimeService() {
+        if (dateTimeService == null) {
+            dateTimeService = SpringContext.getBean(DateTimeService.class);
+        }
+        return dateTimeService;
+    }
+
+    private ConfigurationService getConfigurationService() {
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: Please see above.

File: kfs-purap/src/main/resources/edu/arizona/kfs/module/purap/spring-purap.xml (hunk 1 above)
Comment (adamkost) [reply kfs-purap/src/main/resources/edu/arizona/kfs/module/purap/spring-purap.xml]: As far as I'm aware, you can't change the class and inherit from a parent bean at the same time.

Comment (adamkost) [reply kfs-purap/src/main/resources/edu/arizona/kfs/module/purap/spring-purap.xml]: Adjusted.

File: kfs-purap/src/main/resources/edu/arizona/kfs/module/purap/spring-purap.xml (hunk 2 above)
Comment (adamkost) [reply kfs-purap/src/main/resources/edu/arizona/kfs/module/purap/spring-purap.xml]: Adjusted.

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+        }
+        return configurationService;
+    }
+
+    public B2BPurchaseOrderSciquestServiceImpl() {
+        super();
+    }
+
+    @Override
+    public String getCxml(PurchaseOrderDocument purchaseOrder, String requisitionInitiatorId, String password, ContractManager contractManager, String contractManagerEmail, String vendorDuns) {
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: Why would you change this entire method and not just take what was already in 6 and make the change to use the new b2b property?

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+        cxml.append("  <PurchaseOrder>\n");
+        cxml = writePOHeader(cxml, purchaseOrder, initiator, contractManagerEmail, contractManager, vendorDuns);
+        cxml = writePOLines(cxml, purchaseOrder);
+        cxml.append("  </PurchaseOrder>\n");
+        cxml.append("</PurchaseOrderMessage>");
+        LOG.debug("getCxml(): cXML for po number " + purchaseOrder.getPurapDocumentIdentifier() + ":\n" + cxml.toString());
+        return cxml.toString();
+    }
+
+    protected StringBuilder writeHeader(StringBuilder cXml, String password) {
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: If we copy what was in 6 and change the one line in the getCxml(), this is not needed.

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+        cXml.append("    <Timestamp>").append(date.format(d)).append("T").append(time.format(d)).append("+05:30").append("</Timestamp>\n");
+        cXml.append("    <Authentication>\n");
+        cXml.append("      <Identity>Arizona</Identity>\n"); // this should be paramaterized
+        cXml.append("      <SharedSecret>").append(password).append("</SharedSecret>\n");
+        cXml.append("    </Authentication>\n");
+        cXml.append("  </Header>\n");
+        return cXml;
+    }
+
+    protected StringBuilder writePOHeader(StringBuilder cXml, PurchaseOrderDocument purchaseOrder, Person initiator, String contractManagerEmail, ContractManager contractManager, String vendorDuns) {
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: Please see the writeHeader() above.

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+        cXml.append("      <AccountingDate>" + purchaseOrder.getPurchaseOrderCreateTimestamp().toString() + "</AccountingDate>\n");
+
+        cXml = writeSupplierSection(cXml, purchaseOrder, contractManager, vendorDuns);
+        cXml = writeBillToSection(cXml, purchaseOrder);
+        cXml = writeShipToSection(cXml, purchaseOrder, initiator);
+        cXml.append("    </POHeader>\n");
+        return cXml;
+    }
+
+    protected StringBuilder writeHeaderTaxShippingHandling(StringBuilder cxml, PurchaseOrderDocument po) {
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: I don't see where this method is being used.  Please remove this method.

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+        cxml.append("        <TaxShippingHandling>\n");
+        cxml.append("          <TSHActualCharge>\n");
+        cxml.append("            <Money currency=\"USD\">" + totalTax.toString() + "</Money>\n");
+        cxml.append("          </TSHActualCharge>\n");
+        cxml.append("        </TaxShippingHandling>\n");
+        cxml.append("      </Tax1>\n");
+        return cxml;
+    }
+
+    protected StringBuilder writeSupplierSection(StringBuilder cXml, PurchaseOrderDocument purchaseOrder, ContractManager contractManager, String vendorDuns) {
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: Please see the writeHeader() above.

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+        cXml.append("              <Number>" + bn(phoneNumber) + "</Number>\n");
+        cXml.append("            </TelephoneNumber>\n");
+        cXml.append("          </Phone>\n");
+        cXml.append("        </ContactInfo>\n");
+        cXml.append("      </Supplier>\n");
+
+        return cXml;
+    }
+
+    protected StringBuilder writeBillToSection(StringBuilder cXml, PurchaseOrderDocument purchaseOrder) {
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: Please see the writeHeader() above.

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+        cXml.append("          <State>" + bn(stateCode) + "</State>\n");
+        cXml.append("          <PostalCode>" + bn(postalCode) + "</PostalCode>\n"); // Required.
+        cXml.append("          <Country isocountrycode=\"" + bn(countryIsoCode) + "\">" + bn(countryCode) + "</Country>\n");
+        cXml.append("        </Address>\n");
+        cXml.append("      </BillTo>\n");
+
+        return cXml;
+    }
+
+    protected StringBuilder writeShipToSection(StringBuilder cXml, PurchaseOrderDocument purchaseOrder, Person initiator) {
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: Please see the writeHeader() above.

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+        cXml.append("          <PostalCode>" + bn(postalCode) + "</PostalCode>\n");
+        cXml.append("          <Country isocountrycode=\"" + bn(countryIsoCode) + "\">" + bn(country) + "</Country>\n");
+        cXml.append("        </Address>\n");
+        cXml.append("      </ShipTo>\n");
+
+        return cXml;
+    }
+
+    @SuppressWarnings("unchecked")
+    protected StringBuilder writePOLines(StringBuilder cXml, PurchaseOrderDocument purchaseOrder) {
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: Please see the writeHeader() above.

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+                // pass in whether the POLine is taxable or not
+                cXml.append("      </LineCharges>\n");
+
+                cXml.append("    </POLine>\n");
+            }
+        }
+        return cXml;
+    }
+
+    protected KualiDecimal calcTaxRate(KualiDecimal preTaxAmount, KualiDecimal taxAmount) {
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: This is not needed if we remove the writePOLines() and take what is in baseline KFS6 for getCxml().

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+            return KualiDecimal.ZERO;
+        }
+        return taxAmount.divide(preTaxAmount);
+    }
+
+    /**
+     * Takes in a string that may be null, and guarantees a non-null return.
+     * Returns an empty string if the passed-in value is null.
+     */
+    protected String bn(String value) {
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: This is not needed if we override what is in baseline KFS6 for getCxml().

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,375 @@
...
+            return StringUtils.EMPTY;
+        }
+        return value.trim();
+    }
+
+    /**
+     * Returns true if the system has been configured to use DUNS vendor numbers rather than
+     * traditional internal vendor numbers.
+     */
+    protected boolean isDunsNumberEnabled() {
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: This is not needed if we remove the writeSupplierSection() and take the getCxml() from baseline KFS6.

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,375 @@
...
+        return cxml;
+    }
+
+    protected StringBuilder writeSupplierSection(StringBuilder cXml, PurchaseOrderDocument purchaseOrder, ContractManager contractManager, String vendorDuns) {
+
+        // **START AZ** KFSI-750
+        // determine if the system is configured to use DUNS numbers, rather than VendorNumbers, if so, use that to
+        // filter vendor-specific items off the cart
+        VendorDetail vendor = purchaseOrder.getVendorDetail();
+        String vendorNumberOrDUNS = (isDunsNumberEnabled() ? vendor.getVendorDunsNumber() : vendor.getVendorNumber());
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: Why are we adding the fix for KFSI-750 here?  Do we know if this was fixed in KFS6 baseline getCxml()?

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,375 @@
...
+
+        return cXml;
+    }
+
+    protected StringBuilder writeShipToSection(StringBuilder cXml, PurchaseOrderDocument purchaseOrder, Person initiator) {
+
+        // AddressCode. A code to identify the address, that is sent to the supplier.
+        String addressCode = purchaseOrder.getDeliveryCampusCode() + purchaseOrder.getOrganizationCode();
+        String contactLine1 = StringUtils.isNotBlank(purchaseOrder.getDeliveryBuildingRoomNumber()) ? "Rm " + purchaseOrder.getDeliveryBuildingRoomNumber() : "";
+        // START KFSI-5802
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: Same as the KFSI-750 comment above.  Do we know if this was fixed in KFS6 baseline to know if this is actually needed?

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,375 @@
...
+            contactLine5 = purchaseOrder.getRouteCode();
+        } else if (StringUtils.isNotBlank(purchaseOrder.getDeliveryBuildingName())) {
+            contactLine5 = purchaseOrder.getDeliveryBuildingName();
+        }
+
+        /** *** SHIP TO SECTION **** */
+        cXml.append("      <ShipTo>\n");
+        cXml.append("        <Address>\n");
+        cXml.append("          <TemplateName>Ship To</TemplateName>\n");
+        // cXml.append("          <AddressCode>"+bn(addressCode)+"</AddressCode>\n");
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: If this isn't needed, please remove this comment.

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,375 @@
...
+        String countryCode = purchaseOrder.getBillingCountryName();
+        String countryIsoCode = purchaseOrder.getBillingCountryCode();
+
+        /** *** BILL TO SECTION **** */
+        // Contact - There can be 0-5 Contact elements. The label attribute is optional.
+        // There must be 1-5 AddressLine elements. The label attribute is optional.
+        cXml.append("      <BillTo>\n");
+        cXml.append("        <Address>\n");
+        cXml.append("          <TemplateName>Bill To</TemplateName>\n");
+        // cXml.append("          <AddressCode>"+bn(addressCode)+"</AddressCode>\n");
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: If this is not needed, please remove.

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,375 @@
...
+
+        // check indicator to decide if receiving or delivery address should be sent to the vendor
+        if (!purchaseOrder.getAddressToVendorIndicator()) { // use receiving address
+            addressLine1 = purchaseOrder.getDeliveryBuildingLine1Address();
+            addressLine2 = purchaseOrder.getDeliveryBuildingRoomNumber();
+            // if (StringUtils.isNotBlank(purchaseOrder.getDeliveryBuildingLine2Address())) {
+            // addressLine3 = purchaseOrder.getDeliveryBuildingLine2Address();
+            // } else {
+            // addressLine3 = " ";
+            // }
```
Comment (leahasullivan) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/service/impl/B2BPurchaseOrderSciquestServiceImpl.java]: If the lines above are not needed, please remove.

**Comment by leahasullivan** [comment]
Initial review completed.  Please see comments above and retest.

**Pull request closed without merging**